### PR TITLE
Refactored relational expression translation to more easily support tran...

### PR DESCRIPTION
**Pull request by anpete**
...slation enhancements.

### CODE REVIEW

**Comment by anpete** [comment]
@lukewaters This should help with the coalesce enhancement...

File: src/EntityFramework.Relational/Query/RelationalQueryModelVisitor.cs [hunk 1]
```diff
@@ -223,104 +217,73 @@ var innerJoinExpression
...
-            var projectionCounts
-                = _queriesBySource
-                    .Select(kv => new
-                        {
-                            SelectExpression = kv.Value,
-                            kv.Value.Projection.Count
-                        })
-                    .ToList();
+            var selectExpression = TryGetQuery(queryModel.MainFromClause);
+            var requiresClientEval = selectExpression == null;
```
Comment (lukewaters) [inline src/EntityFramework.Relational/Query/RelationalQueryModelVisitor.cs]: Is there any way that we could get a selectExpression back that still requires partial client evaluation?

**Comment by lukewaters** [comment]
:shipit:

File: src/EntityFramework.Relational/Query/RelationalQueryModelVisitor.cs (hunk 1 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/RelationalQueryModelVisitor.cs]: This is just tracking whether this particular WhereClause requires client eval.

**Pull request merged**